File: src/VitalSynth.Core/Session.cs
namespace VitalSynth.Core;

/// <summary>
/// A running VitalSynth session. Owns the oscillators, slow modulators
/// and the HL7 tick cadence. UI-agnostic — a console, Blazor WASM or
/// WPF client all consume the same session by calling <see cref="Step"/>
/// and reading the exposed vitals.
///
/// The engine is <b>passive</b>: the client chooses when to step and
/// how many samples to produce, which keeps threading, animation and
/// pause semantics simple across platforms.
/// </summary>
public sealed class Session : IDisposable
{
    private readonly HeartOscillator _heart;
    private readonly SlowSignals _slow;
    private readonly IVitalsSink? _sink;
    private readonly int _samplesPerHl7;
    private int _hl7Counter;

    public Cfg Cfg { get; }

    public int    CurrentBpm   => _heart.CurrentBpm;
    public double CurrentSpo2  => _slow.LastSpo2;
    public double CurrentTempC => _slow.LastTempC;

    /// <summary>Raised on every HL7 tick, after the sink has been published to.</summary>
    public event Action<Vitals>? VitalsPublished;

    public Session(Cfg cfg, IVitalsSink? sink = null)
    {
        Cfg = cfg;
        _heart = new HeartOscillator(cfg);
        _slow  = new SlowSignals(cfg);
        _sink  = sink;
        _samplesPerHl7 = Math.Max(1, cfg.SampleRateHz * cfg.Hl7IntervalMs / 1000);

        _slow.AdvanceSpo2(0);
        _slow.AdvanceTemperature(0);
    }

    /// <summary>
    /// Advance the engine by <paramref name="samples"/> samples at the
    /// configured sample rate, filling <paramref name="buffer"/> with
    /// ECG voltages in mV. Triggers HL7 ticks when due.
    /// </summary>
    public void Step(Span<double> buffer)
    {
        var dt = 1.0 / Cfg.SampleRateHz;
        for (var i = 0; i < buffer.Length; i++)
        {
            var resp = _slow.Respiration(dt);
            buffer[i] = _heart.NextSample(dt, resp);
            _slow.AdvanceSpo2(dt);
            _slow.AdvanceTemperature(dt);

            if (++_hl7Counter >= _samplesPerHl7)
            {
                _hl7Counter = 0;
                var v = new Vitals(
                    HeartRate:    _heart.CurrentBpm,
                    Spo2:         _slow.LastSpo2,
                    TemperatureC: _slow.LastTempC,
                    Scenario:     Cfg.Scenario,
                    TimestampUtc: DateTime.UtcNow);
                _sink?.Publish(v);
                VitalsPublished?.Invoke(v);
            }
        }
    }

    public void Dispose()
    {
        if (_sink is IDisposable d) d.Dispose();
    }
}

File: src/VitalSynth.Core.FS/Session.fs
namespace VitalSynth.Core.FS

open System

/// A running VitalSynth session.
///
/// The line `let dt : float<s> = 1.0 / cfg.SampleRate` below is the
/// money shot: `cfg.SampleRate : float<Hz>`, `Hz` is defined as `/s`,
/// so `1.0 / float<Hz>` is automatically inferred as `float<s>`. No
/// runtime check, no manual `.Value` extraction — the compiler does it.
///
/// `VitalsPublished` is typed as `EventHandler<Vitals>` rather than F#'s
/// native `Event<Vitals>` so C# consumers get an idiomatic
/// `session.VitalsPublished += (sender, v) => …` without `FSharpHandler<T>`.
type Session(cfg: Cfg) =
    let heart = HeartOscillator(cfg)
    let slow  = SlowSignals(cfg)
    let samplesPerHl7 =
        max 1 (int (Math.Round(float cfg.SampleRate * cfg.Hl7Interval.TotalSeconds)))
    let mutable hl7Counter = 0
    let mutable pendingBeats = 0
    let vitalsPublished = Event<EventHandler<Vitals>, Vitals>()

    do
        slow.AdvanceSpo2 0.0<s> |> ignore
        slow.AdvanceTemperature 0.0<s> |> ignore

    member _.Cfg = cfg
    member _.CurrentHeartRate = heart.CurrentHeartRate
    member _.CurrentSpo2      = slow.LastSpo2
    member _.CurrentTempC     = slow.LastTemperature
    member _.PendingBeats     = pendingBeats

    [<CLIEvent>]
    member _.VitalsPublished = vitalsPublished.Publish

    member _.ConsumePendingBeats () =
        let n = pendingBeats
        pendingBeats <- 0
        n

    /// Fills <paramref name="buffer"/> with ECG samples in millivolts.
    member this.Step (buffer: float[]) =
        let dt : float<s> = 1.0 / cfg.SampleRate   // <— Hz⁻¹ inferred as s
        for i in 0 .. buffer.Length - 1 do
            let resp = slow.Respiration dt
            let sample = heart.NextSample(dt, resp)
            buffer.[i] <- inMillivolts sample
            if heart.BeatThisSample then pendingBeats <- pendingBeats + 1
            slow.AdvanceSpo2 dt |> ignore
            slow.AdvanceTemperature dt |> ignore

            hl7Counter <- hl7Counter + 1
            if hl7Counter >= samplesPerHl7 then
                hl7Counter <- 0
                let v : Vitals = {
                    HeartRate    = heart.CurrentHeartRate
                    Spo2         = slow.LastSpo2
                    TemperatureC = slow.LastTemperature
                    Scenario     = cfg.Scenario
                    TimestampUtc = DateTime.UtcNow
                }
                vitalsPublished.Trigger(this, v)

    interface IDisposable with
        member _.Dispose () = ()
